### PR TITLE
Finish Validator client 

### PR DESCRIPTION
**Pull request by GregTheGreek**
Should finalize validator duties.

Targets latest master on `ethereum/eth2.0-specs`

### CODE REVIEW

**Comment by GregTheGreek** [comment]
I'd like to belive this is ready to review for a First itteration on the general structure re-work @wemeetagain

File: validator/validator.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  private blockService: BlockProcessingService;
+  private genesisInfo: GenesisInfo;
+  public isActive: boolean;
+
+  /**
+   * @param {ValidatorCtx} ctx
+   */
+  public constructor(ctx: ValidatorCtx) {
+    this.ctx = ctx;
+    this.logger = (msg: string) => console.log(msg);
```
Comment (mpetrunic) [inline validator/validator.ts]: use new logger? It's singleton so you can just import it anywhere

Comment (GregTheGreek) [reply validator/validator.ts]: Loollll I forgot about this thanks!

File: validator/block.ts [hunk 2]
```diff
@@ -5,14 +5,15 @@ import {blsSign} from "../src/stubs/bls";
 import {getDomain, slotToEpoch} from "../src/chain/helpers/stateTransitionHelpers";
 import {DOMAIN_BEACON_BLOCK, DOMAIN_RANDAO} from "./constants";
 import {getEmptyBlock} from "../src/chain/helpers/genesis";
+import {AbstractLogger} from "../src/logger";
```
Comment (mpetrunic) [inline validator/block.ts]:
```javascript
import logger from "../src/logger";


logger.info("log");
```

It's singleton so you can just import and use it, no need to pass it as prop everywhere.

File: validator/validator.ts [hunk 3]
```diff
@@ -51,7 +52,7 @@ class Validator {
    * @returns {Promise<void>}
    */
   public async setup(): Promise<void> {
-    this.logger("Setting up validator client...");
+    this.logger.info.info("Setting up validator client...");
```
Comment (mpetrunic) [inline validator/validator.ts]: that's wierd heh

Comment (GregTheGreek) [reply validator/validator.ts]: find and replace failed me

File: validator/block.ts (hunk 2 above)
Comment (mpetrunic) [reply validator/block.ts]: @GregTheGreek think you missed this comment :)

**Comment by wemeetagain** [comment]
CI tests aren't running for some reason, maybe the merge conflict?

**Comment by GregTheGreek** [comment]
@wemeetagain Should be good to go

File: README.md [hunk 4]
```diff
@@ -27,6 +27,12 @@ You will need to go over the [specification](https://github.com/ethereum/eth2.0-
 1. `yarn install`
 2. `yarn test`
 
+# Building
+1. If you haven't `yarn install`
+2. `yarn build`
+
+Two executables are available, `./bin/lodestar` and `./bin/lodestar-client`
```
Comment (wemeetagain) [inline README.md]: I don't think lodestar-client does anything

File: src/validator/validator.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  /**
+   * Creates a new block proccessing service and starts it.
+   */
+  private async start(): Promise<void> {
+    await this.setup();
+    this.run();
+  }
+
```
Comment (wemeetagain) [inline src/validator/validator.ts]: we probably want a `stop` method, too

File: src/validator/block.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /**
+   * IFF a validator is selected construct a block to propose.
+   * @returns {Promise<void>}
+   */
+  private async buildBlock() {
+    let block: BeaconBlock = getEmptyBlock();
+    block = await this.assembleHeader(block);
+    block = await this.assembleBody(block);
```
Comment (wemeetagain) [inline src/validator/block.ts:27]: probably should return `block`

File: src/validator/block.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  private async assembleBody(block: BeaconBlock): Promise<BeaconBlock> {
+    return block;
+  }
+
+  /**
+   * Main function to start block processing
+   * @returns {Promise<void>}
+   */
+  public async start() {
+    await this.buildBlock();
```
Comment (wemeetagain) [inline src/validator/block.ts]: I don't think we will want to build a block right away, just when called upon.
In fact, this probably doesn't need start/stop methods.

File: src/validator/block.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    this.validatorIndex = index;
+    this.provider = provider;
+    this.privateKey = privateKey;
+  }
+
+  /**
+   * IFF a validator is selected construct a block to propose.
+   * @returns {Promise<void>}
+   */
+  private async buildBlock() {
```
Comment (wemeetagain) [inline src/validator/block.ts]: maybe make this public, or make another way to return a block.

File: src/chain/helpers/validatorStatus.ts [hunk 9]
```diff
@@ -99,3 +101,59 @@ export function prepareValidatorForWithdrawal(state: BeaconState, index: Validat
...
   validator.withdrawalEpoch = getCurrentEpoch(state) + MIN_VALIDATOR_WITHDRAWAL_DELAY;
 }
+
+/**
+ * Return the committee assignment in the ``epoch`` for ``validator_index`` and ``registry_change``.
+ * ``assignment`` returned is a tuple of the following form:
+ * ``assignment[0]`` is the list of validators in the committee
+ * ``assignment[1]`` is the shard to which the committee is assigned
+ * ``assignment[2]`` is the slot at which the committee is assigned
+ * ``assignment[3]`` is a bool signalling if the validator is expected to propose
```
Comment (wemeetagain) [inline src/chain/helpers/validatorStatus.ts]: this line seems stale

File: src/validator/validator.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    this.blockService = new BlockProcessingService(this.validatorIndex, this.provider, this.ctx.privateKey);
+    // TODO setup attestation service
+  }
+
+  private async checkAssignment(): Promise<void> {
+    // If epoch boundary then look up for new assignment
+    if ((Date.now() - this.genesisInfo.startTime) % SLOTS_PER_EPOCH === 0) {
+      const epoch = await this.provider.getCurrentEpoch();
+      const assignment: CommitteeAssignment = await this.provider.getCommitteeAssignment(epoch, this.validatorIndex);
+      const isProposer: boolean = this.provider.isProposerAtSlot(assignment.slot, this.validatorIndex);
```
Comment (wemeetagain) [inline src/validator/validator.ts:120]: it seems like this is relying on the beacon chain to calculate if the validator is a proposer at a given slot.
I think we should have the validator calculate if they're a proposer for a given slot.

File: src/chain/helpers/validatorStatus.ts [hunk 11]
```diff
@@ -99,3 +101,59 @@ export function prepareValidatorForWithdrawal(state: BeaconState, index: Validat
...
+ * a beacon block at the assigned slot.
+ * @param {BeaconState} state
+ * @param {Epoch} epoch
+ * @param {ValidatorIndex} validatorIndex
+ * @param {boolean} registryChange
+ * @returns {{validators: ValidatorIndex[]; shard: Shard; slot: number; isProposer: boolean}}
+ */
+export function getCommitteeAssignment(
```
Comment (wemeetagain) [inline src/chain/helpers/validatorStatus.ts:118]: I think this function (and `isProposerAtSlot`) should go under `src/validator` somewhere.
This is something that the validator should calculate for themselves.

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looking good.

I think the main thing is we'll probably want to move most of the validator processing logic into the validator itself, so the beacon node can remain blissfully unaware of validator concerns. It simply serves data its seen, and broadcasts data its given.
I think this means being able to sync the BeaconState from the beacon node, being able to run the state transition logic within the validator.

Maybe these changes can happen in a future PR though.

File: src/validator/block.ts (hunk 8 above)
Comment (GregTheGreek) [reply src/validator/block.ts]: Yeah this makes more sense. Removing start/stop and making this public actually makes a lot more sense.

File: src/validator/block.ts (hunk 7 above)
Comment (GregTheGreek) [reply src/validator/block.ts]: yeah i agree

File: README.md (hunk 4 above)
Comment (GregTheGreek) [reply README.md]: I think you're right!

File: src/validator/validator.ts (hunk 10 above)
Comment (GregTheGreek) [reply src/validator/validator.ts:120]: Right, but that would mean the validator needs to maintain a pruned copy of the state in order to make that calculation. I can check in on how other clients do these opperations, but I'm curious as to what you think.

IMO thats definitely the next iteration anyway - Need to add a DB to the validator to catch slashing conditions and also that would help facilitate this.

File: src/chain/helpers/validatorStatus.ts (hunk 11 above)
Comment (GregTheGreek) [reply src/chain/helpers/validatorStatus.ts:118]: I'm going to hold off on this until we architect a proper storage mechanism for the client

**Review by wemeetagain — DISMISSED** [review]

**Review by mpetrunic — DISMISSED** [review]

**Comment by GregTheGreek** [comment]
@mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme

**Comment by mpetrunic** [comment]
> @mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme

It only counts reviews from people with write access to repo, so I can't help you anyways :)

**Comment by GregTheGreek** [comment]
> > @mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme
> 
> It only counts reviews from people with write access to repo, so I can't help you anyways :)

Shouldn't be an issue anymore 😄

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**